pytorch_lightning/trainer/trainer.py

```diff
@@ -1494,6 +1500,42 @@ def call_hook(
 
         return output
 
+    def _profile_and_call(self, obj: object, hook_name: str, *args: Any, **kwargs: Any) -> Optional[Any]:
+        fn = getattr(obj, hook_name)
+        with self.profiler.profile(f"{obj.__class__.__name__}.{hook_name}"):
+            return fn(*args, **kwargs)
+
+    def _call_hook(
+        self,
+        obj: object,
+        hook_name: str,
+        *args: Any,
+        pl_module: Optional["pl.LightningModule"] = None,
+        **kwargs: Any,
```

Comment:
i'm confused why `obj` and `pl_module` are passed in here. we should always be able to derive the pl_module from the trainer.

Reply:
for pl_module see @carmocca's answer here: https://github.com/PyTorchLightning/pytorch-lightning/issues/8506#issuecomment-969343535

I agree it would be much cleaner if we could not pass `pl_module`.

for obj, we could avoid passing it by creating the three methods you mentioned above, and each method is called directly, i.e. we wouldn't need `_call_hook`:
`_call_lightning_module_hook`
`_call_callback_hook`
`_call_strategy_hook`

The pros to this approach are the code is easier to read and understand, and less args passed, with a con being some code duplication. I think the three methods approach makes sense

Reply:
I checked and there are only two places where we actually pass in `pl_module` to call_hook:
https://github.com/PyTorchLightning/pytorch-lightning/blob/dcafc95f2b0fd3f176d425139ca99676ce943a12/pytorch_lightning/trainer/optimizers.py#L35
https://github.com/PyTorchLightning/pytorch-lightning/blob/dcafc95f2b0fd3f176d425139ca99676ce943a12/pytorch_lightning/trainer/data_loading.py#L577

Will create an issue for both